(filters_frustum) Add colors to exported ply

csparker247 opened the pull request:
This colors the faces of the frustums exported by `ExportCameraFrustums`. This is meant to be an aid to help with orientating poses visually. 

This PR is currently just basic implementation in order to begin discussing details.

Questions/Thoughts:

1.  Should colors be optional?
2. Is there a better/preferred way to define color constants? I haven't really looked into the other colorization features in openMVG yet. Is `image::RGBColor` a good const class? Does it have an `ostream` operator?
3. I used colors generated by MeshLab. Do we want to tweak these?
4. What will it take to [add axis lines](https://github.com/openMVG/openMVG/issues/1700#issuecomment-602950501)? By this, I think you mean the camera's axes of rotation, centered on the pinhole.

pmoulon commented:
> Should colors be optional?
- Could be.
- I'm ok with default could having one top color different than the side face.

> Is there a better/preferred way to define color constants? I haven't really looked into the other colorization features in openMVG yet. Is image::RGBColor a good const class? Does it have an ostream operator?
- Yes there is color constant defined in image namespace. You are right it could be nice to define those with ostream iterator or having a std::to_string(RGBColor) working on it. The stream operator would have to match the `PLY` requirement `R G B`

> I used colors generated by MeshLab. Do we want to tweak these?
- Your initial suggestion was nice (top face with one color, and all the other with another fixed color)

> What will it take to add axis lines? By this, I think you mean the camera's axes of rotation, centered on the pinhole.
- It would consist in adding Edge to the PLY files (see here -> http://paulbourke.net/dataformats/ply/)
- Edges in PLY are linking some vertex with a Line (I don't know if it is supported by all 3D viewer) but it is a nice way to display 3D lines. I would like to use this feature for a while to display the visibility graph of the sfm_data scene but time is lacking... We could experiment here and see how easy to use it is.

csparker247 commented:
Since there's already an `ostream` operator for `Rgb`, I'm not sure we can go that route unless you don't mind changing the format of that function from `{r, g, b}` to `r g b`. It seems like it's probably best to not mess with it. I'll look into alternatives.

csparker247 commented:
Updated with: 
1. Uses `image::RGBColor` for colors
2. Adds RGB->PLY Color string function within the Filter_Frustum compilation unit.
3. Only the top face is colored red. Others are middle gray.
4. Colorization optional. Updated `ExportCameraFrustums` accordingly.

Still going to look into the axis lines using PLY Edges.

pmoulon commented:
Looking good for me as it is.
I tried and I saw that PLY EDGES are not supported by all the 3DViewer, so let's keep what we have.
Or we could add the axis here https://github.com/openMVG/openMVG/tree/develop/src/software/SfMWebGLViewer

I think the best for the future would be to write our own sfm_data viewer (3D scene) and editor (to edit ID and camera model).

pmoulon commented on src/openMVG/sfm/sfm_data_filters_frustum.cpp:
```diff
@@ -181,26 +192,28 @@ const
 
   // Export frustums faces
   IndexT count = 0;
+  std::string color_top_face = colorize ? RgbPLYString(image::RED) : "";
```
Nit: consider using `const` if we know we not gonna modify it later

pmoulon commented:
Thx you @csparker247 
I cannot wait to work with you on your next contribution.

csparker247 commented:
No problem! Glad to help!
